great_expectations/zep/context.py

```diff
@@ -0,0 +1,149 @@
+from __future__ import annotations
+
+import logging
+from pprint import pformat as pf
+from typing import TYPE_CHECKING, Callable, Dict, List, Optional
+
+from great_expectations.util import camel_to_snake
+from great_expectations.zep.type_lookup import TypeLookup
+
+if TYPE_CHECKING:
+    from great_expectations.zep.interfaces import Datasource
+
+SourceFactoryFn = Callable[..., "Datasource"]
+
+LOGGER = logging.getLogger(__name__)
+
+
+def _remove_suffix(s: str, suffix: str) -> str:
+    # NOTE: str.remove_suffix() added in python 3.9
+    if s.endswith(suffix):
+        s = s[: -len(suffix)]
+    return s
+
+
+def _get_simplified_name_from_type(
+    t: type, suffix_to_remove: Optional[str] = None
+) -> str:
+    result = camel_to_snake(t.__name__)
+    if suffix_to_remove:
+        return _remove_suffix(result, suffix_to_remove)
+    return result
+
+
+class _SourceFactories:
+    """
+    Contains a collection of datasource factory methods in the format `.add_<TYPE_NAME>()`
+
+    Contains a `.type_lookup` dict-like two way mapping between previously registered `Datasource`
+    or `DataAsset` types and a simplified name for those types.
+    """
+
+    type_lookup = TypeLookup()
+    __source_factories: Dict[str, SourceFactoryFn] = {}
+
+    @classmethod
+    def register_factory(
+        cls,
+        ds_type: type,
+        fn: SourceFactoryFn,
+        asset_types: Optional[List[type]] = None,
+    ) -> None:
+        """
+        Add/Register a datasource factory function.
+
+        Derives a SIMPLIFIED_NAME from the provided `Datasource` type.
+        Attaches a method called `add_<SIMPLIFIED_NAME>()`.
+
+        Also registers related `DataAsset` types.
+
+        Example
+        -------
+        `class PandasDatasource` -> `add_pandas()`
+        """
+        simplified_name = _get_simplified_name_from_type(
+            ds_type, suffix_to_remove="_datasource"
+        )
+
+        method_name = f"add_{simplified_name}"
+        LOGGER.info(
+            f"2a. Registering {ds_type.__name__} as {simplified_name} with {method_name}() factory"
+        )
+
+        pre_existing = cls.__source_factories.get(method_name)
+        if pre_existing:
+            raise ValueError(f"{simplified_name} factory already exists")
+
+        # TODO: simplify or extract the following datasource & asset type registration logic
+        asset_types = asset_types or []
+        asset_type_names = [
+            _get_simplified_name_from_type(t, suffix_to_remove="_asset")
+            for t in asset_types
+        ]
+
+        # NOTE: This check is a shortcut. What we need to protect against is different asset types
```

Comment:
I'd make this a TODO. I think we'll need to namespace the assets to datasources since I can see an asset type being used by more than 1 datasource. We can defer that for now but I'd be more explicit here like:
`# TODO: We should namespace the asset type to the datasource so different datasources can reuse asset types.`

Reply:
I'll update this.